### ISSUE TITLE
Proposal: drop-in directories

### ISSUE THREAD

**Issue by squeed** [containernetworking/cni#928]
As mentioned in #927, the the reality is that network configuration is not managed top-down by an administrator in real-world clusters. Rather, it is maintained by one or more third parties, all of which may need to react according to the existing configuration and/or modify it.

This makes composability awkward, to say the least.

An _extremely common_ use-case is wanting to insert a chained plugin in to an existing configuration. That existing configuration is managed by yet another third (fourth?) party. Currently, this can only be done by reading, editing, and re-writing the configuration file. This is fragile, since the file is nominally "owned" by one network provider, but edited by another.

This proposal is for CNI dropin directories, which `libcni` would read from to scan for additional chained plugins.

### Example

Imagine there is a configuration file, `10-awesome.conflist` with the following contents:

```jsonc
{
    "cniVersion": "1.0",
    "name": "my-awesome-network",
    "plugins": [
        { "type": "awesome" },
        { "type": "chained-1" }
    ]
}
```

Then, libcni would look for a folder named `my-awesome-network`. It would parse any `.conf` files contained in them. Imagine it finds `11-chained.conf` with the following contents:

```jsonc
{
    "type": "chained-2"
}
```

This would cause `libcni` to insert "chained-2" to the *end* of list of chained plugins when parsing `10-awesome.conflist`.

**Comment by bleggett** [comment]
> Currently, this can only be done by reading, editing, and re-writing the configuration file. This is fragile, since the file is nominally "owned" by one network provider, but edited by another.

Yes, and if there are multiple node agents installing plugins that are mutually unaware of each other, they essentially are contending over the same node-level config file, which can and does lead to [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) bugs which render the entire node unusable.

It seems simplest and safest to effectively mimic Unix init system design here, which encountered and solved the same problem years ago using the mechanism proposed by @squeed

**Comment by bleggett** [comment]
Took a stab at a quick spec update to add this in #1052 (looking ahead to 1.2+) - mostly to clarify what a spec-based approach would look like.
